src/activities/Elsa.Activities.Timers.Hangfire/Jobs/RunHangfireWorkflowJob.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Threading.Tasks;
+
+using Elsa.Activities.Timers.Hangfire.Models;
+using Elsa.Activities.Timers.Services;
+using Elsa.Models;
+using Elsa.Persistence;
+using Elsa.Services;
+
+using Hangfire;
+
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Elsa.Activities.Timers.Hangfire.Jobs
+{
+    public class RunHangfireWorkflowJob
+    {
+        private readonly IWorkflowRunner _workflowRunner;
+        private readonly IWorkflowRegistry _workflowRegistry;
+        private readonly IWorkflowInstanceStore _workflowInstanceManager;
+        private readonly IServiceProvider _serviceProvider;
+
+        public RunHangfireWorkflowJob(IWorkflowRunner workflowRunner, IWorkflowRegistry workflowRegistry, IWorkflowInstanceStore workflowInstanceStore, IServiceProvider serviceProvider)
+        {
+            _workflowRunner = workflowRunner;
+            _workflowRegistry = workflowRegistry;
+            _workflowInstanceManager = workflowInstanceStore;
+            _serviceProvider = serviceProvider;
+        }
+
+        public async Task ExecuteAsync(RunHangfireWorkflowJobModel data)
+        {
+            var workflowBlueprint = (await _workflowRegistry.GetWorkflowAsync(data.WorkflowDefinitionId, data.TenantId, VersionOptions.Published))!;
+            
+            if(workflowBlueprint == null)
+            {
+                return;
+            }
+
+            WorkflowInstance? workflowInstance = null;
+
+            if (data.WorkflowInstanceId == null)
+            {
+                if (workflowBlueprint.IsSingleton == false || await _workflowInstanceManager.GetWorkflowIsAlreadyExecutingAsync(data.TenantId, data.WorkflowDefinitionId) == false)
+                {
+                    await _workflowRunner.RunWorkflowAsync(workflowBlueprint, data.ActivityId);
+                }
+            }
+            else
+            {
+                for (var i = 0; i < 3 && workflowInstance == null; i++)
```

Comment:
Perhaps extract this retry mechanism to a private method? Makes the `ExecuteAsync` a little bit easier to read. It's OK if you don't; I can do it later.

Reply:
Can do. I wasn't sure you wanted it that way. Because with Quartz it's an infinite loop if the instance doesn't exist.

Reply:
> Because with Quartz it's an infinite loop if the instance doesn't exist.

Indeed. Which is why I like your solution here 👍